search-service/src/main/kotlin/com/egm/stellio/search/web/TemporalApiResponseBuilder.kt

```diff
@@ -0,0 +1,241 @@
+import org.springframework.util.MultiValueMap
+import java.time.ZonedDateTime
+
+typealias CompactedTemporalAttribute = List<Map<String, Any>>
+typealias TemporalValue = List<Any>
+val SIMPLIFIED_TEMPORAL_VALUE_NAMES =
+    listOf("values", "languageMaps", "valueLists", "jsons", "vocabs", "objects", "objectLists")
+
+typealias Range = Pair<ZonedDateTime, ZonedDateTime>
+
+object TemporalApiResponseBuilder {
+    @SuppressWarnings("LongParameterList")
+    fun buildEntitiesTemporalResponse(
+        entities: List<CompactedEntity>,
+        total: Int,
+        resourceUrl: String,
+        query: TemporalEntitiesQuery,
+        requestParams: MultiValueMap<String, String>,
+        mediaType: MediaType,
+        contexts: List<String>,
+        lang: String? = null
+    ): ResponseEntity<String> {
+        val ngsiLdDataRepresentation = parseRepresentations(requestParams, mediaType)
+        lang?.let { ngsiLdDataRepresentation.copy(languageFilter = it) }
+        val successResponse = buildQueryResponse(
+            entities.toFinalRepresentation(ngsiLdDataRepresentation),
+            total,
+            resourceUrl,
+            query.entitiesQuery.paginationQuery,
+            requestParams,
+            mediaType,
+            contexts
+        )
+        val temporalQuery = query.temporalQuery
+        if (temporalQuery.isLastNInsideLimit()) {
+            return successResponse
+        }
+
+        val attributesWhoReachedLimit = getAttributesWhoReachedLimit(entities, query)
+
+        if (attributesWhoReachedLimit.isEmpty()) {
+            return successResponse
+        }
+
+        val range = getTemporalPaginationRange(attributesWhoReachedLimit, query)
+
+        val filteredEntities = entities.map { filterEntityInRange(it, range, query) }
+        return ResponseEntity.status(HttpStatus.PARTIAL_CONTENT).apply {
+            this.headers(
+                successResponse.headers
+            )
+            this.header(
+                HttpHeaders.CONTENT_RANGE,
+                getHeaderRange(range, query.temporalQuery)
+            )
+        }.body(serializeObject(filteredEntities.toFinalRepresentation(ngsiLdDataRepresentation)))
+    }
+
+    fun buildEntityTemporalResponse(
+        entity: CompactedEntity,
+        query: TemporalEntitiesQuery,
+        mediaType: MediaType,
+        requestParams: MultiValueMap<String, String>,
+        contexts: List<String>,
+    ): ResponseEntity<String> {
+        val ngsiLdDataRepresentation = parseRepresentations(requestParams, mediaType)
+        val successResponseHeader = prepareGetSuccessResponseHeaders(mediaType, contexts)
+
+        if (query.temporalQuery.isLastNInsideLimit()) {
+            return successResponseHeader.body(
+                serializeObject(
+                    entity.toFinalRepresentation(ngsiLdDataRepresentation)
+                )
+            )
+        }
+
+        val attributesWhoReachedLimit = getAttributesWhoReachedLimit(listOf(entity), query)
```

Comment:
I am wondering if it would not be a lot easier to do the range calculations, instances truncations, ... in query service (for instance, using the `temporalEntityAttributesWithMatchingInstances` in https://github.com/stellio-hub/stellio-context-broker/blob/develop/search-service/src/main/kotlin/com/egm/stellio/search/service/QueryService.kt#L72)

Reply:
@thomasBousselin probably the most important of all my comments